### PR TITLE
import: modernizations, and various additions

### PR DESCRIPTION
**Pull request by poettering**
This is split out of #20156, and preparatory work for it.

This isn't particularly well split out, but I still hope it's somewhat OK to review.

The existing tests should cover most of the rafactoring here already, the sysupdate PR will then add tests that validate the new features added.

### CODE REVIEW

File: src/test/test-install-file.c [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+#include "fileio.h"
+#include "install-file.h"
+#include "path-util.h"
+#include "rm-rf.h"
+#include "tests.h"
+#include "tmpfile-util.h"
+#include "umask-util.h"
+
+static void test_install_file(void) {
+
```
Comment (bluca) [inline src/test/test-install-file.c]: extra new line? also `log_info("/* %s */", __func__);`

File: src/import/import-common.h [hunk 2]
```diff
@@ -3,11 +3,19 @@
...
+        IMPORT_READ_ONLY      = 1 << 1, /* make generated image read-only */
+        IMPORT_BTRFS_SUBVOL   = 1 << 2, /* tar: preferably create images as btrfs subvols */
+        IMPORT_BTRFS_QUOTA    = 1 << 3, /* tar: set up btrfs quota for new subvolume as child of parent subvolume */
+        IMPORT_CONVERT_QCOW2  = 1 << 4, /* raw: if we detect a qcow2 image, unpack it */
+        IMPORT_DIRECT         = 1 << 5, /* import without rename games */
+        IMPORT_SYNC           = 1 << 6, /* fsync() right before we are done */
 
-        IMPORT_FLAGS_MASK = IMPORT_FORCE|IMPORT_READ_ONLY,
+        IMPORT_FLAGS_MASK_TAR = IMPORT_FORCE|IMPORT_READ_ONLY|IMPORT_BTRFS_SUBVOL|IMPORT_BTRFS_QUOTA|IMPORT_DIRECT|IMPORT_SYNC,
+        IMPORT_FLAGS_MASK_RAW = IMPORT_FORCE|IMPORT_READ_ONLY|IMPORT_CONVERT_QCOW2|IMPORT_DIRECT|IMPORT_SYNC,
```
Comment (bluca) [inline src/import/import-common.h:18]: INVAL/MAX ?

File: src/import/import-tar.c [hunk 3]
```diff
@@ -175,56 +175,75 @@ static int tar_import_finish(TarImport *i) {
...
-        r = btrfs_subvol_make_fallback(i->temp_path, 0755);
+        if (FLAGS_SET(i->flags, IMPORT_DIRECT|IMPORT_FORCE))
+                (void) rm_rf(d, REMOVE_ROOT|REMOVE_PHYSICAL|REMOVE_SUBVOLUME);
+
+        if (i->flags & IMPORT_BTRFS_SUBVOL)
+                r = btrfs_subvol_make_fallback(d, 0755);
+        else
+                r = mkdir(d, 0755) < 0 ? -errno : 0;
+        if (r == -EEXIST && (i->flags & IMPORT_DIRECT)) /* EEXIST is OK if in direct mode, but not otherwise,
+                                                       * because in that case our temporary path collided */
```
Comment (bluca) [inline src/import/import-tar.c]: nit: looks mis-aligned

File: src/import/import-common.h (hunk 2 above)
Comment (yuwata) [reply src/import/import-common.h:18]: We do not introduce _INVALID and _MAX for bit field enum.

**Comment by bluca** [comment]
is it worth thinking about multiple hash formats support? Debian packages already do SHA512, hard-coding a particular hash format historically has often not been the best in hindsight

File: src/import/pull-common.c [hunk 4]
```diff
@@ -237,45 +212,82 @@ int pull_make_auxiliary_job(
...
+                *ret_checksum_job = *ret_signature_job = NULL;
+                return 0;
+        }
+
+        r = import_url_last_component(url, &fn);
+        if (r < 0 && r != -EADDRNOTAVAIL) /* EADDRNOTAVAIL means there was no last component, which is OK for
+                                           * us, we'll just assume it's not a checksum/signature file */
+                return r;
+
+        /* Acquire the checksum file if verification or sinature verification is requested and the main file
```
Comment (bluca) [inline src/import/pull-common.c]: /sinature/signature/

File: src/import/pull-common.c [hunk 5]
```diff
@@ -594,3 +652,28 @@ int pull_job_restart_with_sha256sum(PullJob *j, char **ret) {
...
+
+        return hostname_is_valid(name, 0);
+}
+
+int pull_url_needs_checksum(const char *url) {
+        _cleanup_free_ char *fn = NULL;
+        int r;
+
+        /* Returns true if we need to validate this resource via a hash value. This returns true for all
+         * files — except for gpg signature files and SHA256SUMS files and the like, which are validated a
```
Comment (bluca) [inline src/import/pull-common.c]: "with a"

File: src/import/pull-common.h [hunk 6]
```diff
@@ -7,27 +7,31 @@
...
-        PULL_SETTINGS           = 1 << 1,  /* .nspawn settings file */
-        PULL_ROOTHASH           = 1 << 2,  /* only for raw: .roothash file for verity */
-        PULL_ROOTHASH_SIGNATURE = 1 << 3,  /* only for raw: .roothash.p7s file for verity */
-        PULL_VERITY             = 1 << 4,  /* only for raw: .verity file for verity */
+        PULL_FORCE              = 1 << 0, /* replace existing image */
+        PULL_READ_ONLY          = 1 << 1, /* make generated image read-only */
+        PULL_SETTINGS           = 1 << 1, /* download .nspawn settings file */
+        PULL_ROOTHASH           = 1 << 2, /* only for raw: download .roothash file for verity */
+        PULL_ROOTHASH_SIGNATURE = 1 << 3, /* only for raw: download .roothash.p7s file for verity */
+        PULL_VERITY             = 1 << 4, /* only for raw: download .verity file for verity */
```
Comment (bluca) [inline src/import/pull-common.h:15]: I know it's existing stuff, but does it make sense to pull the verity file without the root hash or viceversa?

File: src/import/import-common.h (hunk 2 above)
Comment (poettering) [reply src/import/import-common.h:18]: yes, @yuwata is right, for bit masks _INVALID typically makes no sense. _MAX might, but would probably have to be defined as "all bits on", but that's a bit nasty, and shouldn't be done unless we really need it

File: src/import/pull-common.h (hunk 6 above)
Comment (poettering) [reply src/import/pull-common.h:15]: roothash without verity definitely makes sense: after all verity might be embedded in the main image (i.e. it's what I think is actually the best approach).

verity without roothash might make sense too, for example, if you have the roothash from some other source already, where it was pinned, and you just need the resources that match it.

**Comment by poettering** [comment]
> is it worth thinking about multiple hash formats support? Debian packages already do SHA512, hard-coding a particular hash format historically has often not been the best in hindsight

Wel, this is supposed to be a bit opinionated. i.e. I think it's generally understood that one of the failures of gpg for example is that it still still supports way too many algorithms, 80% of which are not up to today's starndards.

But yes, eventually we'll have to add more hashes, but we should do that when the time comes I think, and only add what is actually standard by then. But if we do this, then this needs some extra work, since then we have to look for SHA512SUMS/SHA256SUMS.gpg file upstream, instead of SHA256SUMS/SHA256SUMS.gpg, so the changes will be a bit bigger.

But yeah, sooner or later we will have to add more variants here, on the top of my list is actually to support .p7s sigs instead of gpg here (or even signify/minisign sigs), so it's very well on the roadmap to add more, but we should only add what really makes sense.

Or to say this differently: at this point in time SHA512 is a candidate to add, but SHA1 or MD5 is definitely not anymore, hence I think we should not make this generic, but decide each time what to add (and then have individual PRs for each)

**Comment by poettering** [comment]
force pushed new version, with all points addresses, though see comments above

**Comment by bluca** [comment]
> force pushed new version, with all points addresses, though see comments above

Thanks - note that I haven't managed to go through everything, it's quite a large diff :-) Will try to finish by end of week

File: src/import/pull-tar.c [hunk 7]
```diff
@@ -221,9 +231,38 @@ static int tar_pull_make_local_copy(TarPull *i) {
         if (!i->local)
                 return 0;
 
-        r = pull_make_local_copy(i->final_path, i->image_root, i->local, i->flags);
+        assert(i->final_path);
+
+        p = strjoina(i->image_root, "/", i->local);
```
Comment (bluca) [inline src/import/pull-tar.c]: prefix_roota?

File: src/import/pull-tar.c [hunk 8]
```diff
@@ -221,9 +231,38 @@ static int tar_pull_make_local_copy(TarPull *i) {
...
+
+        r = tempfn_random(p, NULL, &t);
         if (r < 0)
-                return r;
+                return log_error_errno(r, "Failed to generate temporary filename for %s: %m", p);
+
+        if (i->flags & PULL_BTRFS_SUBVOL) {
```
Comment (bluca) [inline src/import/pull-tar.c]: single line so bracket not needed?

File: src/import/pull-tar.c [hunk 9]
```diff
@@ -306,19 +359,14 @@ static void tar_pull_job_on_finished(PullJob *j) {
                                                             * in per-directory verification mode, since only
                                                             * then the signature is detached, and thus a file
                                                             * of its own. */
-                        log_error_errno(j->error, "Failed to retrieve signature file, cannot verify. (Try --verify=no?)");
-                        r = i->signature_job->error;
+                        r = log_error_errno(i->signature_job->error,
+                                            "Failed to retrieve signature file, cannot verify. (Try --verify=no?)");
```
Comment (bluca) [inline src/import/pull-tar.c:363]: Is actively suggesting to disable validation a good idea? It needs to be available ofc, but perhaps it should not be encouraged

File: src/import/pull-tar.c [hunk 10]
```diff
@@ -346,59 +395,92 @@ static void tar_pull_job_on_finished(PullJob *j) {
...
+                        i->temp_path = mfree(i->temp_path);
+
+                        if (i->settings_job &&
+                            i->settings_job->error == 0) {
+
+                                /* Also move the settings file into place, if it exists. Note that we do so only if we also
+                                 * moved the tar file in place, to keep things strictly in sync. */
+                                assert(i->settings_temp_path);
+
+                                /* Regenerate final name for this auxiliary file, we might know the etag of the file now, and
```
Comment (bluca) [inline src/import/pull-tar.c:456]: etag?

File: src/import/pull.c [hunk 11]
```diff
@@ -203,23 +238,39 @@ static int pull_raw(int argc, char *argv[], void *userdata) {
...
+               "\n%2$sCommands:%3$s\n"
+               "  tar URL [NAME]              Download a TAR image\n"
+               "  raw URL [NAME]              Download a RAW image\n"
+               "\n%2$sOptions:%3$s\n"
                "  -h --help                   Show this help\n"
                "     --version                Show package version\n"
                "     --force                  Force creation of image\n"
                "     --verify=MODE            Verify downloaded image, one of: 'no',\n"
-               "                              'checksum', 'signature'\n"
+               "                              'checksum', 'signature' or SHA256 hash\n"
```
Comment (bluca) [inline src/import/pull.c]: maybe "or a literal SHA256 hash" ? While reading this I thought it meant the "SHA256" string was the accepted value

File: src/import/import-fs.c [hunk 12]
```diff
@@ -286,6 +330,31 @@ static int parse_argv(int argc, char *argv[]) {
                         arg_read_only = true;
                         break;
 
+                case ARG_DIRECT:
+                        arg_direct = true;
```
Comment (bluca) [inline src/import/import-fs.c:334]: why is this a flag while the other are booleans?

File: src/import/import-fs.c [hunk 13]
```diff
@@ -103,44 +100,60 @@ static int progress_bytes(uint64_t nbytes, void *userdata) {
...
+                local = empty_or_dash_to_null(argv[2]);
+        else if (path) {
+                r = path_extract_filename(path, &l);
+                if (r < 0)
+                        return log_error_errno(r, "Failed to extract filename from path '%s': %m", path);
+
+                local = l;
+        }
+
+        if (arg_direct) {
```
Comment (bluca) [inline src/import/import-fs.c:121]: commit message says: "This commit is a bit simpler though, as a --direct mode does not make
sense here" but --direct is supported, maybe reword the commit message to be clearer saying that it's just implementing something different/simpler?

File: docs/ENVIRONMENT.md [hunk 14]
```diff
@@ -316,3 +316,22 @@ fuzzers:
 
 Note that is may be also useful to set `$SYSTEMD_LOG_LEVEL`, since all logging
 is suppressed by default.
+
+systemd-importd:
+
+* `SYSTEMD_IMPORT_BTRFS_SUBVOL` – takes a boolean, which controls whether to
+  prefer creating btrfs subvolumes over plain directories for machine
+  images. Has not effect on non-btrfs file systems where subvolumes are not
```
Comment (bluca) [inline docs/ENVIRONMENT.md]: /Has not effect/Has no effect/

**Review by bluca — APPROVED** [review]
LGTM, just some minor nits.

File: src/import/pull-tar.c (hunk 9 above)
Comment (poettering) [reply src/import/pull-tar.c:363]: Well, I put it in question form and in brackets, to distance myself from this idea. But reality is that it's extremely useful, so I fiugred i might be something to help people with.

And there are OK usecases (for example, if you know the hash beforehand and pass it in, or if you validate externally or so. I think being too "judgy" is not helpful.

I mean I added this originally myself since I kept forgetting what the right switch was after not looking at the code for a bit, so this is primarily there to help myself ;-)

File: src/import/pull-tar.c (hunk 10 above)
Comment (poettering) [reply src/import/pull-tar.c:456]: HTTP etag:

https://en.wikipedia.org/wiki/HTTP_ETag

this is an http client, so context should be clear enough

File: src/import/import-fs.c (hunk 12 above)
Comment (poettering) [reply src/import/import-fs.c:334]: i guess i could also make it a boolean. But my thinking was, that the booleans tweak details about the output, but they don't alter the operation as a whole. the direct mode is different: it changes the operation executed to something quite different. it's like an adjective to the verb that turns the verb into something related but sufficiently different.

not sure if that is enough to make it not a bool...

Comment (bluca) [reply src/import/import-fs.c:334]: that sounds fine, was just wondering

File: src/import/pull-tar.c (hunk 9 above)
Comment (bluca) [reply src/import/pull-tar.c:363]: makes sense

**Comment by poettering** [comment]
Forgot to say: i fixed the issues @bluca found. since minor upgrading green label
